pkg/controller/job/job_controller.go

```diff
@@ -1017,7 +1031,7 @@ func (jm *Controller) trackJobStatusAndRemoveFinalizers(ctx context.Context, job
 			continue
 		}
 		considerPodFailed := isPodFailed(pod, jobCtx.job)
-		if podutil.IsPodTerminal(pod) || considerPodFailed || jobCtx.finishedCondition != nil || jobCtx.job.DeletionTimestamp != nil {
+		if canRemoveFinalizer(logger, jobCtx, pod, considerPodFailed) {
```

Comment:
It seems that we are decoupling removing finalizers from `uncountedTerminatedPods`.

I don't think we should do that. We should only add UIDs to uncountedTerminatedPods when we are ready to remove finalizers. This way, `uncountedTerminatedPods` is only used instantaneously, instead of staying populated for a long time.

Reply:
Done, PTAL.